sources/Engine.ts

```diff
@@ -69,17 +70,19 @@ export class Engine {
       // Ignore errors; too bad
     }
 
-    if (typeof lastKnownGood !== `object` || lastKnownGood === null)
-      return definition.default;
-
-    if (!Object.prototype.hasOwnProperty.call(lastKnownGood, packageManager))
-      return definition.default;
+    if (typeof lastKnownGood === `object` && lastKnownGood !== null &&
+        Object.prototype.hasOwnProperty.call(lastKnownGood, packageManager)) {
+      const override = (lastKnownGood as any)[packageManager];
+      if (typeof override === `string`) {
+        return override;
+      }
+    }
 
-    const override = (lastKnownGood as any)[packageManager];
-    if (typeof override !== `string`)
+    if (process.env.COREPACK_NO_LOOKUP)
       return definition.default;
 
-    return override;
+    const latest = await fetchAsJson(`https://registry.npmjs.org/${packageManager}`);
+    return latest[`dist-tags`].latest;
```

Comment:
This fetch would be executed at every invocation, it'd be quite slow for things like `yarn run some-script` which themselves call other scripts.

I think it'd need to be cached at least on a by-process basis (through an env variable that would be set when spawning the subprocess, for example `COREPACK_LKG_YARN=1.22.18`), and possibly in a system-wide cache (for example to only fetch the new versions every 24 hours).

Reply:
Wouldn't the last downloaded version be cached by `lastKnownGood`?